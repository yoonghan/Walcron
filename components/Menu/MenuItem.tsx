`use strict`

import * as React from "react";
import {fontFamily} from "../../shared/style";
import {BLACK, WHITE, GREY, DARKGREY} from "../../shared/style";

interface MenuItemProps {
  whiteOnBlack?: boolean;
  menuTriggerSize: number;
  childNumber: number;
  totalChild: number;
}

export const MenuItem: React.SFC<MenuItemProps> = ({menuTriggerSize, whiteOnBlack, childNumber, totalChild, ...props}) => {
  return (
    <React.Fragment>
      <li>
        {props.children}
      </li>
      <style jsx>{`
        li :global(a) {
          color: ${whiteOnBlack ?  WHITE: BLACK};
          font-family: ${fontFamily.standard};
          text-transform: uppercase;
          font-weight: 600;
          letter-spacing: -0.02em;
          word-wrap: break-word;
          user-select: none;
          z-index: 1000;
        }

        li :global(a):hover {
          color: ${whiteOnBlack ?  GREY: DARKGREY};
        }

        @media screen and (max-width: ${menuTriggerSize}px) {
          :global(.show) > li {
            transition-delay: ${0.04 * (childNumber + 1)}s;
            z-index: ${totalChild - (childNumber + 1)}
          }

          li {
            opacity: 0;
            transform: translate(0, -10px) scale(1.07);
            will-change: transform, opacity;
            transition: opacity 0.25s,transform 0.2s cubic-bezier(0.165, 0.84, 0.44, 1),scale 0.2s cubic-bezier(0.165, 0.84, 0.44, 1);
          }

          :global(.show) > li {
            opacity: 1;
            transform: translate(0, 0) scale(1);
          }

          li :global(a) {
            display: block;
            padding: 1rem 2rem;
            position: relative;
            color: ${WHITE};
            font-size: 1rem;
            text-transform: none;
            background: ${BLACK};
            margin-bottom: -1px;
            transition: color 0.12s cubic-bezier(0.455, 0.03, 0.515, 0.955);
          }

          li :global(a):hover {
            color: ${GREY};
          }
        }
        `}</style>
    </React.Fragment>
  );
}
