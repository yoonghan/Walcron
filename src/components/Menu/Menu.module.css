.container {
  background-color: var(--menu-bg);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  display: flex;
}

.container .command {
  flex: 1;
}

.container .menu {
  flex: 1;
}

.container .wrapper {
  display: flex;
  justify-content: center;
  padding: 0.5rem 2rem;
}

.container .menu img {
  padding-right: 1rem;
}

.container .menu a {
  white-space: nowrap;
}

.container .divider {
  padding: 0 1rem;
}

.container .switchButton {
  padding: 1rem;
  color: var(--button);
  background: var(--menu-bg);
  font-weight: bold;
  cursor: pointer;
  border: 0;
  user-select: none;
}

@media only screen and (max-width: 480px) {
  .container .menu {
    padding: 0.5rem 1rem;
    font-size: 0.75rem;
  }

  .container .menu img {
    width: 15px;
  }

  .container .divider {
    padding: 0 0.5rem;
  }
}
