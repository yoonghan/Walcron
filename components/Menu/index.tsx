`use strict`

/**
  Menu is created with performance consideration where state is not used instead using classList remove/add.
  **/

import * as React from "react";
import {BLACK, WHITE, GREY, DARKGREY} from "../../shared/style";
import {fontFamily} from "../../shared/style";
import {MenuItem} from "./MenuItem";

export interface MenuProps {
  whiteOnBlack?: boolean;
  menuItems: Array<JSX.Element>;
}

export default class Menu extends React.PureComponent<MenuProps, {}> {
  private menuTriggerSize = 767;
  private menuRef = React.createRef<HTMLUListElement>();
  private mobileMenuRef = React.createRef<HTMLSpanElement>();
  private documentBackgroundElement?:HTMLElement;

  constructor(props:MenuProps) {
    super(props);
  };

  _drawMenuItem = (menuItem:JSX.Element, index:number) => {
    return <MenuItem
      key={"menu_menuItem_" + index}
      whiteOnBlack={this.props.whiteOnBlack}
      menuTriggerSize={this.menuTriggerSize}
      childNumber={index}
      totalChild={this.props.menuItems.length}
    >{menuItem}</MenuItem>;
  }

  _drawMenus = () => {
    return this.props.menuItems.map(this._drawMenuItem);
  }

  _click_Menu = () => {
    if(this.menuRef.current) {
      if(!this.menuRef.current.classList.contains("show")) {
        this.menuRef.current.classList.add("show");
        document.body.classList.add("mobile-modal");
      }
      else {
        this.menuRef.current.classList.remove("show");
        document.body.classList.remove("mobile-modal");
      }
    }
  }

  _click_doc_layout = (e:Event) => {
    const target:HTMLElement = e.target as HTMLElement;
    if(this.mobileMenuRef.current && !this.mobileMenuRef.current.contains(target) &&
        this.menuRef.current && this.menuRef.current.classList.contains("show")){
      this._click_Menu();
    }
  }

  componentWillMount() {
    if ((process as any).browser){
      document.addEventListener('click', this._click_doc_layout, false);
      this.documentBackgroundElement = document.createElement("DIV");
      this.documentBackgroundElement.className = "menu-background";
      document.body.appendChild(this.documentBackgroundElement);
    }
  }

  componentWillUnmount() {
    if ((process as any).browser){
      document.removeEventListener('click', this._click_doc_layout, false);
      if(this.documentBackgroundElement) {
        document.body.removeChild(this.documentBackgroundElement);
      }
    }
  }

  render() {
    const {whiteOnBlack} = this.props;
    return (
      <React.Fragment>
        <nav className={"menu"}>
          <span className={"nav-toggle"} onClick={this._click_Menu} ref={this.mobileMenuRef}>
            Menu
          </span>
          <ul ref={this.menuRef}>
            {this._drawMenus()}
          </ul>
        </nav>
        <style jsx global>{`
          @media screen and (max-width: ${this.menuTriggerSize}px){
            body.mobile-modal .menu-background {
              position: absolute; top: 0; bottom: 0; right: 0; left: 0; background: rgba(0,0,0,0.3); z-index: 999
            }
          }
          `}
        </style>
        <style jsx>{`
          .menu {
            background: ${whiteOnBlack ? 'rgba(0,0,0,0.4)': 'none'};
          }
          .menu a, .menu span {
            text-decoration: none;
            color: ${whiteOnBlack ?  WHITE: BLACK};
            font-family: ${fontFamily.standard};
            text-transform: uppercase;
            font-weight: 600;
            letter-spacing: -0.02em;
            word-wrap: break-word;
            user-select: none;
            z-index: 1000;
          }
          .menu a:hover {
            color: ${whiteOnBlack ?  GREY: DARKGREY};
          }
          .menu .nav-toggle {
            display: none;
          }
          .menu ul {
            list-style-type: none;
            margin: 0;
            padding: 0;
            display: flex;
            justify-content: space-between;
            align-items: center;
            z-index: 1000;
          }
          @media screen and (max-width: ${this.menuTriggerSize}px) {
            .menu {
              background-color: ${BLACK};
              color: ${WHITE};
            }

            .menu .nav-toggle {
              display: inline-block;
              cursor: pointer;
              position: relative;
            }

            .menu ul {
              flex: 0 1 100%;
              max-height: 0;
              opacity: 0;
              position: relative;
              overflow: hidden;
            }

            .menu ul.show {
              display: block;
              opacity: 1;
              overflow: visible;
            }

            .menu a, .menu span {
              color: ${WHITE};
            }
          }
        `}</style>

      </React.Fragment>
    );
  }
}
