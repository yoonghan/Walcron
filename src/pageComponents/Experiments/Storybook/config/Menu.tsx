import Menu from "@/components/Menu"

const pages = [
  {
    path: "/sample1",
    display: "Sample 1",
  },
  {
    path: "/sample2",
    display: "Sample 2",
  },
]

const SampleMenu = () => <Menu pages={pages} />

export default SampleMenu
