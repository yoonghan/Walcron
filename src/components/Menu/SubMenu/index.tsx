import { findAllChildByPath } from "@/config/pages"
import { WithRouterProps } from "next/dist/client/with-router"
import Link from "next/link"
import { withRouter } from "next/router"
import React from "react"
import style from "./SubMenu.module.css"

export interface Props extends WithRouterProps {}

const SubMenu = ({ router }: Props) => {
  const pathname = router.pathname
  const allChild = findAllChildByPath(pathname)

  const isCurrentPath = (menuPath: string) => pathname == menuPath

  if (allChild.length === 0) {
    return <div data-testid="empty"></div>
  } else
    return (
      <div className={style.container}>
        {allChild.map((menu, index) => (
          <React.Fragment key={menu.path}>
            {isCurrentPath(menu.path) ? (
              <i className={style.divider}>{menu.display}</i>
            ) : (
              <Link href={menu.path} className={style.divider}>
                {menu.display}
              </Link>
            )}
          </React.Fragment>
        ))}
      </div>
    )
}

export default withRouter(SubMenu)
