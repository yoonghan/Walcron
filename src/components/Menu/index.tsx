/* eslint-disable @next/next/no-html-link-for-pages */
/* eslint-disable @next/next/no-img-element */
import { sortedMenuPagesWithFilteredHomeAndSubMenu } from "@/config/pages"
import { animated, useChain, useSpring, useSpringRef } from "@react-spring/web"
import { WithRouterProps } from "next/dist/client/with-router"
import Link from "next/link"
import { withRouter } from "next/router"
import React, { useCallback, useState } from "react"
import CommandBar from "../CommandBar"
import style from "./Menu.module.css"
import SubMenu from "./SubMenu"

export interface Props extends WithRouterProps {}

enum Display {
  Menu,
  Command,
}

const Menu = ({ router }: Props) => {
  const pathname = router.pathname
  const [_display, setDisplay] = useState(Display.Menu)

  const menuRef = useSpringRef()
  const [springMenu, apiMenu] = useSpring(() => ({
    ref: menuRef,
    from: { scale: 1, width: "100%" },
  }))
  const commandRef = useSpringRef()
  const [springCommand, apiCommand] = useSpring(() => ({
    ref: commandRef,
    from: { scale: 0, width: "0", height: "0" },
  }))
  useChain([menuRef, commandRef])
  const [springButton, apiButton] = useSpring(() => ({
    from: { transform: "rotate(0deg)" },
  }))
  useChain([menuRef, commandRef])

  const isNotLastRecord = (index: number) => {
    return index !== sortedMenuPagesWithFilteredHomeAndSubMenu?.length - 1
  }

  const isCurrentPath = (menuPath: string) => pathname == menuPath

  const onSwitchClick = useCallback(() => {
    setDisplay((currentDisplay) => {
      if (currentDisplay === Display.Menu) {
        apiButton.start({
          to: { transform: "rotate(180deg)" },
        })
        apiMenu.start({
          to: { scale: 0, width: "0" },
        })
        apiCommand.start({
          to: { scale: 1, width: "100%", height: "5rem" },
        })
        return Display.Command
      } else {
        apiMenu.start({
          to: { scale: 1, width: "100%" },
        })
        apiCommand.start({
          to: { scale: 0, width: "0%", height: "0" },
        })
        apiButton.start({
          to: { transform: "rotate(0deg)" },
        })
        return Display.Menu
      }
    })
  }, [apiButton, apiCommand, apiMenu])

  return (
    <div className={style.container}>
      <animated.div className={style.menu} style={{ ...springMenu }}>
        <div className={style.wrapper}>
          <a href="/">
            <img src="/img/logo/logo-color.svg" alt="home" width={18} />
          </a>
          {sortedMenuPagesWithFilteredHomeAndSubMenu.map((menu, index) => (
            <React.Fragment key={menu.path}>
              {isCurrentPath(menu.path) ? (
                <i>{menu.display}</i>
              ) : (
                <Link href={menu.path}>{menu.display}</Link>
              )}
              {isNotLastRecord(index) && (
                <span className={style.divider}>|</span>
              )}
            </React.Fragment>
          ))}
        </div>
        <SubMenu />
      </animated.div>
      <animated.div className={style.command} style={{ ...springCommand }}>
        <CommandBar />
      </animated.div>
      <animated.button
        onClick={onSwitchClick}
        className={style.switchButton}
        style={{ ...springButton }}
      >
        ‚å©
      </animated.button>
    </div>
  )
}

export default withRouter(Menu)
