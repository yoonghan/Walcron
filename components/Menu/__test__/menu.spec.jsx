import * as React from 'react';
import { mount } from 'enzyme';
import Menu from "../index";

const menuItems = [
  <React.Fragment/>,
  <div></div>,
  <div>Non Empty</div>
]

describe('Menu', () => {
  it('renders whiteOnBlack normally', () => {
    const wrapper = mount(<Menu whiteOnBlack={true} menuItems={menuItems}/>);
    expect(wrapper.debug()).toMatchSnapshot();
  }),
  it('renders blackOnWhite normally', () => {
    const wrapper = mount(<Menu whiteOnBlack={false} menuItems={menuItems}/>);
    expect(wrapper.debug()).toMatchSnapshot();
  })
});
