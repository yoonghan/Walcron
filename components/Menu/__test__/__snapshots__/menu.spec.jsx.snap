// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders blackOnWhite normally 1`] = `
"<Menu whiteOnBlack={false} menuItems={{...}}>
  <nav className=\\"jsx-2870282083 jsx-1630941244 menu\\">
    <span onClick={[Function]} className=\\"jsx-2870282083 jsx-1630941244 nav-toggle\\">
      Menu
    </span>
    <ul className=\\"jsx-2870282083 jsx-1630941244\\">
      <MenuItem whiteOnBlack={false} menuTriggerSize={767} childNumber={0} totalChild={3}>
        <li className=\\"jsx-781183891\\" />
        <JSXStyle id=\\"3877431110\\" dynamic={{...}} />
      </MenuItem>
      <MenuItem whiteOnBlack={false} menuTriggerSize={767} childNumber={1} totalChild={3}>
        <li className=\\"jsx-809278972\\">
          <div />
        </li>
        <JSXStyle id=\\"3877431110\\" dynamic={{...}} />
      </MenuItem>
      <MenuItem whiteOnBlack={false} menuTriggerSize={767} childNumber={2} totalChild={3}>
        <li className=\\"jsx-3559909334\\">
          <div>
            Non Empty
          </div>
        </li>
        <JSXStyle id=\\"3877431110\\" dynamic={{...}} />
      </MenuItem>
    </ul>
  </nav>
  <JSXStyle id=\\"2870282083\\" />
  <JSXStyle id=\\"2618150123\\" dynamic={{...}} />
</Menu>"
`;

exports[`Menu renders whiteOnBlack normally 1`] = `
"<Menu whiteOnBlack={true} menuItems={{...}}>
  <nav className=\\"jsx-2870282083 jsx-2111747771 menu\\">
    <span onClick={[Function]} className=\\"jsx-2870282083 jsx-2111747771 nav-toggle\\">
      Menu
    </span>
    <ul className=\\"jsx-2870282083 jsx-2111747771\\">
      <MenuItem whiteOnBlack={true} menuTriggerSize={767} childNumber={0} totalChild={3}>
        <li className=\\"jsx-2185469919\\" />
        <JSXStyle id=\\"3877431110\\" dynamic={{...}} />
      </MenuItem>
      <MenuItem whiteOnBlack={true} menuTriggerSize={767} childNumber={1} totalChild={3}>
        <li className=\\"jsx-4004353712\\">
          <div />
        </li>
        <JSXStyle id=\\"3877431110\\" dynamic={{...}} />
      </MenuItem>
      <MenuItem whiteOnBlack={true} menuTriggerSize={767} childNumber={2} totalChild={3}>
        <li className=\\"jsx-154878362\\">
          <div>
            Non Empty
          </div>
        </li>
        <JSXStyle id=\\"3877431110\\" dynamic={{...}} />
      </MenuItem>
    </ul>
  </nav>
  <JSXStyle id=\\"2870282083\\" />
  <JSXStyle id=\\"2618150123\\" dynamic={{...}} />
</Menu>"
`;
